More Tensor Mechanics to automatic differentiation conversion

dschwen opened the pull request:
This adds finite and incremnetal small strain, multiple inelastic stress (with return mapping stuff), power law creep, non-cartesian coordinate systems (RZ and RSPHERICAL materials and kernels).

If this gets merged I have a Bison PR ready as well for a creep model.

Refs #12650 

P.S.: I had to make a cut somewhere as this was growing too much, so the plane strain calculators are written, but retracted in the last commit and will be part of a future PR

Ping @tophmatthews

dschwen committed:
Add strain bases and non-cartesian kernels (#12650)

dschwen committed:
Simplify Compute2DSmallStrain (#12650)

dschwen committed:
Add and fix tests (#12650)

dschwen committed:
Add docs (#12650)

dschwen committed:
Add ADComputeMultipleInelasticStress (#12650)

dschwen committed:
Add single var return mapping solution (#12650)

dschwen committed:
Add RankForTensor macros to MooseTypes.h (#12650)

dschwen committed:
getMaterial for ADMaterials (#12650)

dschwen committed:
Cleanup TM materials (#12650)

dschwen committed:
Foundations for return mapping (#12650)

dschwen committed:
Add non cartesian strain calculators (#12650)

dschwen committed:
Retract 1D plane strain calculators (#12650)

tophmatthews commented:
Dude...epic. I'm going to have to send some green chile your way!!

lindsayad reviewed (commented):
I'm going to look into some of relatively high `relative_tols`

dschwen commented on modules/tensor_mechanics/doc/content/source/kernels/ADStressDivergenceRSphericalTensors.md:
```diff
@@ -0,0 +1,15 @@
+
+!syntax description /ADKernels/ADStressDivergenceRSphericalTensors<RESIDUAL>
+
+ADStressDivergenceRSphericalTensors is the automatic differentiation version of
```
I could just swap this around and make the other file

```
StressDivergenceRSphericalTensors ist the non-automatic differentiation
version of [ADStressDivergenceRSphericalTensors](ADStressDivergenceRSphericalTensors.md).
```

:-D

lindsayad replied:
I'll let @friedmud weigh-in on that. I would be fine with it :laughing:

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/289990/) on de4841a wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12890/site)

This comment will be updated on new commits.

lindsayad commented on modules/tensor_mechanics/test/tests/ad_elastic/tests:
```diff
@@ -0,0 +1,234 @@
+    input = 'rspherical_finite_elastic.i'
+    run_sim = 'True'
+    prereq = 'rspherical_finite_elastic'
+    ratio_tol = 7e-2
```
Yea, there is definitely something wrong here...8 iterations with Newton, only 5 with PJFNK

lindsayad replied:
It's definitely coming from displacement effects in the kernel because if I set `use_displaced_mesh = false`, then the Jacobian is perfect. So it appears to be my problem :smile:

lindsayad replied:
Ok, actually I'm going to say the problem is you and me :smile: So the error comes from:
<pre>
  return _grad_test[_i][_qp](0) * _stress[_qp](0, 0) +               // stress_{rr} part 1
         +(_test[_i][_qp] / <b>_q_point[_qp](0))</b> * _stress[_qp](1, 1) + // stress_{\theta \theta}
         +(_test[_i][_qp] / <b>_q_point[_qp](0))</b> * _stress[_qp](2, 2);  // stress_{\phi \phi}
</pre>
The location of the quadrature point actually changes with the displacements! So what we will need is the volume equivalent of `Assembly::_ad_q_points_face`, e.g. `Assembly::_ad_q_points`. I will get that coded up early next week unless you want to do it

lindsayad replied:
So the Jacobian difference was not just due to the `_q_point`. If I comment out that code, I still get a 10% Jacobian difference...I'm going to have to investigate more...

dschwen commented:
Quite the build fail there... works on my machine with clang though. GCC issue? I'll investigate.

dschwen committed:
gcc dbg compile and non-unity build fix (#12650)

dschwen commented:
Ok, a `using` declaration under the `private` access qualifier will taint parent class member access with this strictest access qualifier down the inheritance chain. Apparently this hets only triggered with GCC in dbg (not GCC opt, clang opt, or clang dbg).

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/290257/) on 9027697 wanted to post the following:

View code coverage site [here](http://mooseframework.inl.gov/docs/PRs/12890/coverage)

dschwen commented:
@lindsayad or @roystgnr I'm getting an FPE

```
Program received signal SIGFPE, Arithmetic exception.
0x00007ffff7dc74f3 in std::sqrt<double, MetaPhysicL::NumberArray<50ul, double> > (in=...) at /home/daniel/Programs/INL/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumber.h:419
419	DualNumber_std_unary(sqrt, 1 / (2 * funcval),)
```

at

```
#1  ADRadialReturnStressUpdate<(ComputeStage)1>::updateState (this=this@entry=0xa0ce00, strain_increment=..., inelastic_strain_increment=..., stress_new=..., elasticity_tensor=..., elastic_strain_old=...)
    at /home/daniel/Programs/INL/moose/modules/tensor_mechanics/src/materials/ADRadialReturnStressUpdate.C:77
77	  ADReal effective_trial_stress = std::sqrt(3.0 / 2.0 * dev_trial_stress_squared);
(gdb) print dev_trial_stress_squared
$1 = {<MetaPhysicL::safe_bool<MetaPhysicL::DualNumber<double, MetaPhysicL::NumberArray<50, double> > >> = {<MetaPhysicL::safe_bool_base> = {<No data fields>}, <No data fields>}, _val = 0, _deriv = {_data = {0 <repeats 50 times>}}}```

where `dev_trial_stress_squared` is zero. I strongly feel that I should be allowed to take a square root of zero without causing an FPU meltdown...

Am I completely out of line to blame MetaPhysicL here? :-)

dschwen committed:
Workaround for dual number FPE in sqrt(0) (#12650)

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/290274/) on e8921d0 wanted to post the following:

View code coverage site [here](http://mooseframework.inl.gov/docs/PRs/12890/coverage)

lindsayad commented:
> I strongly feel that I should be allowed to take a square root of zero without causing an FPU meltdown...

What do you want to happen? When approaching 0 from the positive number side, the derivative tends towards infinity and that's what we assign as the result. You want us to assign an incorrect derivative? Perhaps what you really want is just to not signal a floating point exception?

dschwen commented:
Mhh, right, it does, but the square root of zero is such a common operation it cannot fail. Not throwing an FPE is fine. This is tricky when the pole should cancel out in an expression (something that cause me concern with FParser as well).

dschwen committed:
skip alt PETSc tests (#12650)

dschwen committed:
Address comments (#12650)

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/290325/) on 5e1ff1c wanted to post the following:

View code coverage site [here](http://mooseframework.inl.gov/docs/PRs/12890/coverage)

lindsayad commented:
> but the square root of zero is such a common operation it cannot fail.

What did you do for the old hand-coded case when presumably you've hand-coded the correct analytical Jacobian and have the same issue?

dschwen committed:
Rework docs according to suggestions (#12650)

lindsayad committed:
Add _ad_q_point and _ad_coord

lindsayad committed:
Now tighten Jacobian test tolerances

lindsayad committed:
Mimic libMesh affine behavior

It seems like our function is not the same if we call compute_affine_map or call
compute_single_point_map for each quadrature point. This seems like it might be
a libMesh bug...but my goal in this PR is to mimic the function behavior in the
Jacobian calculation, not ensure that the function evaluation is correct

dschwen commented:
> What did you do for the old hand-coded case when presumably you've hand-coded the correct analytical Jacobian and have the same issue?

I suppose this line, that @tophmatthews added answers this:
```
    if (MooseUtils::absoluteFuzzyEqual(scalar_effective_inelastic_strain, 0.0))
      tangent_operator.zero();
```

I think my workaround in 349f0e4fb1e724fc612086fc9ea8cce41cf2fc49 does effectively the same thing (I should reword that commit message though...).

dschwen commented:
Wooot! With your commits on top it cracks the 7000 line mark.

lindsayad committed:
Add framework tests for new _ad_coord capability

lindsayad reviewed (dismissed):
@dschwen has done an awesome job here porting over one of our most important modules to AD. It's pretty awesome to see perfect Jacobians for finite strain calculations, although I know he wants to run some truly large deformation problems in order to tax the Jacobian calculations further

I can't merge this, however, since I now have important commits on this branch

lindsayad commented:
> I should reword that commit message though

You mean the message where you accuse `MetaPhysicL` of being broken? 😄

roystgnr commented:
> You mean the message where you accuse `MetaPhysicL` of being broken?

DualNumber sqrt(0), specifically, which is why it deserves rewording.  If the accusation was merely a generic "MetaPhysicL is broken *somewhere*" then ... yeah, probably, and apologies in advance for whenever it bites you.  Maybe keep any future use of the compile-time-sparsity-pattern classes in an "experimental don't design any reactors with this" branch for a while...

dschwen commented:
Yeah that one! Sorry! It looks like temporarily disabling (or trapping) FPEs is not feasible (too slow). And I presume you don't want sqrt to lie about its derivative. I guess the workaround is adding the check for a zero argument where ever we take the square root? On the other hand I wonder how useful the current behavior is. What is gained by the FPE, or by setting a derivative to _inf_?

What about this case:

```
ADReal f(ADReal a) 
{
  return std::sqrt(a) * std::sqrt(a);
}
```

The return value should have a derivative of 1, but MetaPhysicL will calculate it as 2*inf*0... whatever that is. I think there is no value in populating a derivative with 1/0 as it does not carry with it the right side limit information required to make it useful in further operations.

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/291040/) on 1117c1b wanted to post the following:

View code coverage site [here](http://mooseframework.inl.gov/docs/PRs/12890/coverage)

roystgnr commented:
> I guess the workaround is adding the check for a zero argument where ever we take the square root?

Check for zero or use a non-AD Real anywhere you take a square root and the value might be zero and you know the derivative effectively never gets used.

> On the other hand I wonder how useful the current behavior is. What is gained by the FPE, or by setting a derivative to _inf_?

Either way you avoid propagating an incorrect value.  Having ```sqrt(x) * sqrt(x)``` give a derivative of NaN is annoying, but at least you then see it and can replace the code with a workaround.  If we instead set the ```sqrt(0)``` derivative to be anything other than inf or nan, that would mean ```sqrt(0) * sqrt(0)``` would have a derivative evaluate as 0, not 1.

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/291303/) on a8acd0b wanted to post the following:

View code coverage site [here](http://mooseframework.inl.gov/docs/PRs/12890/coverage)

tophmatthews commented:
I'm always a little afraid of missing something while using these advanced code pieces. I could see either making it clear about the pit falls with AD in the documentation, or what about using a method that automatically checked for a zero?
```
ADReal psqrt(ADReal a) 
{
  if (a <= 0)
    mooseError("Hey you, protect against zero or negative values when taking the square root")
  return std::sqrt(a);
}
```

dschwen commented:
I don't want it to error out on zero. We compute norms of zero tensors and vectors all the time. It'd be awkward to check if all entries are zero or split each norm into a "norm squared", a test for zero, and a sqrt part.

dschwen commented:
@lindsayad , the `ad_plastic` tests still have a tolerance of 5e-6. I have commits ready for AD BCs (PresetBC, FunctionPresetBC, and Pressure), but it still does not drop the error for some non-linear iterations.

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/common/seealsoADStressDivergenceKernels.md:
```diff
@@ -0,0 +1,5 @@
+## The stress divergence family of automatic differentiation kernels
+
+- [ADStressDivergenceTensors](/ADStressDivergenceTensors.md) - Stress divergence kernel for cartesian coordinates `Problem/coord_type=XYZ` (default)
```
```suggestion
- [ADStressDivergenceTensors](/ADStressDivergenceTensors.md) - Stress divergence kernel for Cartesian coordinates `Problem/coord_type=XYZ` (default)
```

bwspenc committed:
Update modules/tensor_mechanics/doc/content/modules/tensor_mechanics/common/seealsoADStressDivergenceKernels.md

Co-Authored-By: dschwen <daniel@schwen.de>

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/292175/) on 0a2c1eb wanted to post the following:

View code coverage site [here](http://mooseframework.inl.gov/docs/PRs/12890/coverage)

lindsayad commented:
Ok I think I will be conscious enough tomorrow to take a look at the plastic tests. Yay for tamiflu!

bwspenc commented:
I've looked through this, and this looks good to me. I think we should go ahead and merge this despite the slightly loose tolerances on the ad_plastic tests. We can keep investigating that, but they seem pretty good to me.

permcody commented:
@dschwen - This fails 2 tests in distributed mesh mode.

dschwen commented:
Yeah, I just noticed it myself:
https://civet.inl.gov/job/292266/

They may be impacted by @lindsayad's framework changes in this PR. I haven't touched them.

lindsayad commented:
To wrap up the Jacobian bit...I think everything looks really good. On my Mac the largest Jacobian ratios look to be about 1e-6 and these always correspond to the zeroth non-linear iteration where the system state is at its most "wacky", e.g. a state most likely to cause finite-differencing headaches. If we see larger ratios somewhere down the line and if we ever run into cases where we see less non-linear iterations with PJFNK than with NEWTON, then that would be a cause for concern

tophmatthews commented on modules/tensor_mechanics/include/materials/ADPowerLawCreepStressUpdate.h:
```diff
@@ -0,0 +1,77 @@
+  const bool _has_temp;
+
+  /// Temperature variable value
+  const VariableValue & _temperature;
```
@dschwen Isn't this supposed to be ADVariableValue?
